src/utils/AssayGeneratorMetabo.cpp

```diff
@@ -0,0 +1,523 @@
+
+    registerDoubleOption_("transition_threshold", "<num>", 10, "Further transitions need at least x% of the maximum intensity (default 10%)", false);
+
+    registerFlag_("use_known_unknowns", "Use features without identification", false);
+
+  }
+
+  // map precursors to closest feature and retrieve annotated metadata (if possible)
+  // extract meta information from featureXML (MetaboliteAdductDecharger)
+  map<const BaseFeature*, vector<size_t>> extractMetaInformation(const PeakMap & spectra, const KDTreeFeatureMaps& fp_map_kd, const double& precursor_mz_tolerance, const double& precursor_rt_tolerance)
+  {
+    map<const BaseFeature*, vector<size_t>> map_feature_to_precursors;
+
+    // map precursors to closest feature and retrieve annotated metadata (if possible)
+    for (size_t index = 0; index != spectra.size(); ++index)
+    {
+      if (spectra[index].getMSLevel() != 2) { continue; }
+
+      // get precursor meta data (m/z, rt)
+      const vector<Precursor> & pcs = spectra[index].getPrecursors();
+
+      if (!pcs.empty())
+      {
+        const double mz = pcs[0].getMZ();
+        const double rt = spectra[index].getRT();
+
+        // query features in tolerance window
+        vector<Size> matches;
+
+        fp_map_kd.queryRegion(rt - precursor_rt_tolerance, rt + precursor_rt_tolerance, mz - precursor_mz_tolerance, mz + precursor_mz_tolerance, matches, true);
+
+        // no precursor matches the feature information found
+        if (matches.empty()) { continue; }
+
+        // in the case of multiple features in tolerance window, select the one closest in m/z to the precursor
+        Size min_distance_feature_index(0);
+        double min_distance(1e11);
+        for (auto const & k_idx : matches)
+        {
+          const double f_mz = fp_map_kd.mz(k_idx);
+          const double distance = fabs(f_mz - mz);
+          if (distance < min_distance)
+          {
+            min_distance = distance;
+            min_distance_feature_index = k_idx;
+          }
+        }
+        const BaseFeature* min_distance_feature = fp_map_kd.feature(min_distance_feature_index);
+
+        map_feature_to_precursors[min_distance_feature].push_back(index);
+      }
+    }
+    return map_feature_to_precursors;
+  }
+
+  // precursor correction (highest intensity)
+  Int getHighestIntensityPeakInMZRange(double test_mz, const MSSpectrum& spectrum1, double left_tolerance, double right_tolerance)
+  {
+    MSSpectrum::ConstIterator left = spectrum1.MZBegin(test_mz - left_tolerance);
+    MSSpectrum::ConstIterator right = spectrum1.MZEnd(test_mz + right_tolerance);
+
+    // no MS1 precursor peak in +- tolerance window found
+    if (left == right || left->getMZ() > test_mz + right_tolerance)
+    {
+      return -1;
+    }
+
+    MSSpectrum::ConstIterator max_intensity_it = std::max_element(left, right, Peak1D::IntensityLess());
```

Comment:
are the input data expected to be centroided? this may not be the best approach for profile data

Reply:
The input data is expected to be centroided (will add an error if it is profile data.)